drivers/input/inport/pnp.c

```diff
@@ -0,0 +1,285 @@
+
+                break;
+            }
+
+            default:
+                DPRINT("[%p:%X:%X] Unrecognized resource type %X\n",
+                       Descriptor,
+                       Descriptor->ShareDisposition,
+                       Descriptor->Flags,
+                       Descriptor->Type);
+                break;
+        }
+    }
+
+    if (!FoundBasePort || !FoundIrq)
+    {
+        DPRINT1("The device resources were not found\n");
+        Status = STATUS_DEVICE_CONFIGURATION_ERROR;
+        goto Complete;
+    }
+
+    DPRINT("I/O base at %p\n", DeviceExtension->IoBase);
+    DPRINT("IRQ %d\n", RawVector);
+
+    Status = InPortWmiRegistration(DeviceExtension);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("WMI registration failed 0x%X\n", Status);
+        goto Complete;
+    }
+
+    InPortInitializeMouse(DeviceExtension);
+
+    Status = IoConnectInterrupt(&DeviceExtension->InterruptObject,
+                                InportIsr,
+                                DeviceExtension,
+                                NULL,
+                                DeviceExtension->InterruptVector,
+                                DeviceExtension->InterruptLevel,
+                                DeviceExtension->InterruptLevel,
+                                DeviceExtension->InterruptMode,
+                                DeviceExtension->InterruptShared,
+                                DeviceExtension->InterruptAffinity,
+                                FALSE);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Could not connect to interrupt %d\n", DeviceExtension->InterruptVector);
+        InPortWmiDeRegistration(DeviceExtension);
+        goto Complete;
+    }
+
+    OldIrql = KeAcquireInterruptSpinLock(DeviceExtension->InterruptObject);
+    InPortStartMouse(DeviceExtension);
+    KeReleaseInterruptSpinLock(DeviceExtension->InterruptObject, OldIrql);
+
+    DeviceExtension->State = dsStarted;
+
+Complete:
+    MmUnlockPagableImageSection(SectionHandle);
+
+    Irp->IoStatus.Status = Status;
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+InPortRemoveDevice(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp)
+{
+    NTSTATUS Status;
+    LARGE_INTEGER Delay;
+    PINPORT_DEVICE_EXTENSION DeviceExtension = DeviceObject->DeviceExtension;
+
+    PAGED_CODE();
+
+    /* 0.5 seconds, should be enough */
+    Delay.QuadPart = -5000000;
+
+    InPortWmiDeRegistration(DeviceExtension);
+
+    DeviceExtension->State = dsRemoved;
+
+    Irp->IoStatus.Status = STATUS_SUCCESS;
+    IoSkipCurrentIrpStackLocation(Irp);
+    Status = IoCallDriver(DeviceExtension->Ldo, Irp);
+
+    IoReleaseRemoveLockAndWait(&DeviceExtension->RemoveLock, Irp);
+
+    if (DeviceExtension->InterruptObject)
+        IoDisconnectInterrupt(DeviceExtension->InterruptObject);
+
+    /* Wait for DPC */
+    KeDelayExecutionThread(KernelMode, FALSE, &Delay);
+
+    InPortStopMouse(DeviceExtension);
```

Comment:
You should tell the device to stop issuing interrupts before calling `IoDisconnectInterrupt`, not after.